### PR TITLE
[WIP] RunImageJ

### PR DESCRIPTION
**Pull request by mcquin**
Run an ImageJ plugin from CellProfiler.

## Installation
Installation requires [Miniconda](https://conda.io/miniconda.html).

You will also need to clone the [ImageJ server repository](https://github.com/imagej/imagej-server).

**OS X and Linux users:** Edit `environment.yml` to include the `bjornfjohansson` channel and the `wxpython=3.0.2.0` dependency.

Then run the following commands:

```
cd cellprofiler-plugins
conda env create -f environment.yml
source activate cellprofiler-plugins
```

## Use
Start the ImageJ server:

```
cd imagej-server
mvn -Pexec
```

Launch CellProfiler:

```
cd cellprofiler-plugins
pythonw -m cellprofiler --plugins-directory .
```

/cc @ctrueden

### CODE REVIEW

**Comment by mcquin** [comment]
Status update:

I added a "Connect" button to the module GUI that, when pressed, will try to establish a connection to the ImageJ server:

screen shot 2018-03-09 at 1 02 59 pm

If the connection cannot be established, a validation warning appears (as shown above). A validation error will be raised on pipeline load or run if a connection with the configured host cannot be established.

Once the connection is established (note: this does require the user clicking "Connect"), then the ImageJ module dropdown appears and its inputs/outputs can be populated:

screen shot 2018-03-09 at 1 05 25 pm

A connection will automatically be established when a pipeline is loaded. This also makes it possible to run the pipeline headless. 

Remaining TODOs are captured in code comments.

File: runimagej.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        for setting, value in zip(self.input_settings.settings, setting_values[offset:offset + n_input_settings]):
+            setting.value = value
+
+        for setting, value in zip(self.output_settings.settings, setting_values[offset + n_input_settings:]):
+            setting.value = value
+
+    def run(self, workspace):
+        # FIXME: Keep the original IDs in some data structure,
+        # so that we guarantee we run the correct module here.
+        id = self._ij.find(self.ij_module.value)[0]
```
Comment (AetherUnbound) [inline runimagej.py]: This shadows the builtin python `id` function, it might be a good idea to rename this.

File: runimagej.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+        # FOR DEBUGGING
+        logger.debug(json.dumps(details, indent=4))
+
+        for input_ in details["inputs"]:
+            name = input_["name"]
+            input_["rawType"] = raw_type = input_["genericType"].split("<")[0].split(" ")[-1]
+
+            # HACK: For now, we skip service and context parameters.
+            # Later, the ImageJ Server will filter these out for us.
```
Comment (AetherUnbound) [inline runimagej.py]: If there's a GitHub issue for ImageJ or something to link here, that would be excellent!

Comment (ctrueden) [reply runimagej.py]: Indeed—the issue to follow is imagej/imagej-server#14.

**Comment by ctrueden** [comment]
@AetherUnbound @0x00b1 Anything I can do to help move this PR forward? I could put my student @kkangle on tidying up any loose ends, perhaps? What do you think?

We will also want to update this module once imagej/imagej-server#14 is merged and released.

**Pull request merged**